map_creator/src/find_base_byAverage.cpp

```diff
@@ -0,0 +1,358 @@
+//Creating a new grasp pose in the origin co-ordinate
+    vector<double> new_grasp_pos;
+    tf2::Vector3 new_grasp_vec;
+    tf2::Quaternion new_grasp_quat;
+    new_grasp_vec =  new_grasp_trns.getOrigin();
+    new_grasp_quat =  new_grasp_trns.getRotation();
+    new_grasp_quat.normalize();
+    new_grasp_pos.push_back(new_grasp_vec[0]);
+    new_grasp_pos.push_back(new_grasp_vec[1]);
+    new_grasp_pos.push_back(new_grasp_vec[2]);
+    new_grasp_pos.push_back(new_grasp_quat[0]);
+    new_grasp_pos.push_back(new_grasp_quat[1]);
+    new_grasp_pos.push_back(new_grasp_quat[2]);
+    new_grasp_pos.push_back(new_grasp_quat[3]);
+
+//Check the new grasp_pose for Ik
+    Kinematics k;
+    std::vector<double> joints;
+    
+    joints.resize(6);
+    if(k.isIKSuccess(new_grasp_pos, joints, numOfSolns))
+      return true;
+    else
+      return false;
+
+} 
+
+bool areQuaternionClose(tf2::Quaternion q1, tf2::Quaternion q2)
+{
+     double dot = q1.dot(q2);
+     if(dot<0)
+	return false;
+     else
+	return true;
+}
+
+tf2::Quaternion inverseSignQuaternion(tf2::Quaternion q)
+{
+    tf2::Quaternion q_inv(-q[0], -q[1], -q[2], -q[3]);
+    return q_inv;
+}
+
+void findOptimalPosebyAverage(const vector<geometry_msgs::Pose> probBasePoses,geometry_msgs::Pose& final_base_pose){
+//This Function has been borrowed from http://wiki.unity3d.com/index.php/Averaging_Quaternions_and_Vectors
+
+    double totalVecX, totalVecY, totalVecZ;
+    double avgVecX, avgVecY, avgVecZ;  
+    vector<tf2::Quaternion> quatCol;
+    for(int i=0;i<probBasePoses.size();++i)
+    {
+      totalVecX += probBasePoses[i].position.x;
+      totalVecY += probBasePoses[i].position.y;
+      totalVecZ += probBasePoses[i].position.z;
+      tf2::Quaternion quats(probBasePoses[i].orientation.x, probBasePoses[i].orientation.y, probBasePoses[i].orientation.z, probBasePoses[i].orientation.w);
+      quatCol.push_back(quats);
+    }
+   
+    avgVecX = totalVecX/double(probBasePoses.size());
+    avgVecY = totalVecY/double(probBasePoses.size());
+    avgVecZ = totalVecZ/double(probBasePoses.size());
+    
+   
+    
+    double totalQuatX, totalQuatY, totalQuatZ, totalQuatW;
+    double avgQuatX, avgQuatY, avgQuatZ, avgQuatW;
+    for(int j=0;j<quatCol.size();++j)
+    {
+      if(!areQuaternionClose(quatCol[0],quatCol[j]))  
+      {
+	
+        tf2::Quaternion quat_new;
+        quat_new = inverseSignQuaternion(quatCol[j]);
+       
+        totalQuatX +=quat_new[0];
+        totalQuatY +=quat_new[1];
+        totalQuatZ +=quat_new[2];
+        totalQuatW +=quat_new[3];
+        }
+      else
+      {
+	totalQuatX +=quatCol[j][0];
+        totalQuatY +=quatCol[j][1];
+        totalQuatZ +=quatCol[j][2];
+        totalQuatW +=quatCol[j][3];
+      }
+    }
+    avgQuatX = totalQuatX/double(probBasePoses.size());
+    avgQuatY = totalQuatY/double(probBasePoses.size());
+    avgQuatZ = totalQuatZ/double(probBasePoses.size());
+    avgQuatW = totalQuatW/double(probBasePoses.size());
```

Comment:
I don't think you can average quaternions the way you can average other numbers.